src/syntaxtree.c

```diff
@@ -0,0 +1,1072 @@
+/****************************************************************************
+**
+*W  syntaxtree.c
+**
+**
+** TODO:
+**  - Expressions vs Statements
+**  - Documentation
+**  - Compiler functions per argument (will make more separate functions go away)
+**  - Floats
+*/
+#include "system.h"
+#include <stdarg.h>
+
+#include "code.h"
+#include "objects.h"
+
+#include "exprs.h"
+#include "gapstate.h"
+#include "stats.h"
+
+#include "compiled.h"
+#include "vars.h"
+
+#include <ctype.h>
+
+typedef UInt4 LVar;
+typedef UInt4 HVar;
+typedef UInt  GVar;
+
+typedef Obj (*CompileFuncT)(Obj node, Expr expr);
+typedef Obj (*CompileArgT)(Expr expr);
+
+typedef struct {
+    const Char *argname;
+    CompileArgT argcomp;
+} ArgT;
+
+typedef struct {
+    UInt         tnum;
+    CompileFuncT compile;
+    const Char * name;
+    UInt         arity;
+    ArgT         args[6]; /* This is not really a restriction */
+} CompilerT;
+
+static const CompilerT StatCompilers[];
+static const CompilerT ExprCompilers[];
+
+#define COMPILER_ARITY(...)                                                  \
+    (sizeof((ArgT []){ __VA_ARGS__ }) / sizeof(ArgT))
```

Comment:
Unfortunately this macro (or rather, its expansion) trips up HPC-GAP.

No idea how hard it would be to teach ward about this C99 feature. But of course a simple fix would be to forgo variadics, and simply have several `COMPILER` macros: for 0, 1, 2, ... args. But that's certainly less elegant :-/